### ISSUE TITLE
Shipping API: Setting Validation not working

### ISSUE THREAD

**Issue by bilfeldt** [woocommerce/woocommerce#21484]
**Describe the bug**
Shipping Method plugins use the well-documented [Settings API](https://docs.woocommerce.com/document/settings-api/). Unfortunately then validating the fields setup using `init_form_fields()` fails to show any error introduced by throwing an exception. The setting which fails is not actually updated, but no error is shown to the user.

This error was probably introduced in version 3.4.0, and is discussed in a [Github commit](https://github.com/woocommerce/woocommerce/commit/057c87c457c6d4deb83a7d85f93655707ec94eea#commitcomment-30315059) as well.

**To Reproduce**
1. Install a very basic shipping plugin consisting of just one file `wc-skeleton-shipping-method-example/wc-skeleton-shipping-method-example.php`. The code for the file is taken from the official Shipping Plugin Example from WooCommerce, which can be seen [here](https://docs.woocommerce.com/document/shipping-method-api/#section-6) with a simple validation function. The entire code is attached at the bottom of this post.
2. Click on 'WooCommerce->Settings->Shipping->Your Shipping Method'
3. Enter '10' in the Test settings-field and press 'Save'
4. Now you will see that the setting is not saved, but instead of an error appearing in the top saying "Invalid value: 10", then you are seeing two errors saying "TestError: Key:test::Value:10"

**Screenshots**
Here is a screenshot of the result:
<img width="795" alt="screen shot 2018-10-03 at 00 05 08" src="https://user-images.githubusercontent.com/30228807/46379722-0671fa00-c6a0-11e8-8e13-ee54fab0156a.png">

**Expected behavior**
Instead of seeing two error messages saying "TestError: Key:test::Value:10", then we expected to see one error message saying "Invalid value: 10", which was caused by the exception and one error message saying "TestError: Key:test::Value:10", which is caused by manually adding an error with `WC_Admin_Settings::add_error("TestError: Key:{$key}::Value:{$value}");`

**Example plugin to reproduce the bug**
This example Shipping Plugin is simple the officielt plugin with an added form field and validation of this field:
```
/**
* Initialize integration settings form fields.
*
* @return void
*/
public function init_form_fields() {
   $log_path = SS_SHIPPING_WC()->get_log_url();
   $agents_address_format = SS_SHIPPING_WC()->get_agents_address_format();

   $this->form_fields = array(
   	'test'            	=> array(
   		'title'           	=> __( 'Test setting' ),
   		'type'            	=> 'text',
   		'default'           => '',
   		'description'     	=> __( 'Test description' ),
   		'desc_tip'        	=> false
   	)
   );
}

public function validate_test_field($key, $value ) {
	WC_Admin_Settings::add_error("TestError: Key:{$key}::Value:{$value}");//Used for debugging
	throw new Exception('Invalid value: {$value}');
}
```
the entire results can be seen here:
```
<?php
/*
Plugin Name: Your Shipping plugin
Plugin URI: https://woocommerce.com/
Description: Your shipping method plugin
Version: 1.0.0
Author: WooThemes
Author URI: https://woocommerce.com/
*/

/**
 * Check if WooCommerce is active
 */
if ( in_array( 'woocommerce/woocommerce.php', apply_filters( 'active_plugins', get_option( 'active_plugins' ) ) ) ) {

	function your_shipping_method_init() {
		if ( ! class_exists( 'WC_Your_Shipping_Method' ) ) {
			class WC_Your_Shipping_Method extends WC_Shipping_Method {
				/**
				 * Constructor for your shipping class
				 *
				 * @access public
				 * @return void
				 */
				public function __construct() {
					$this->id                 = 'your_shipping_method'; // Id for your shipping method. Should be uunique.
					$this->method_title       = __( 'Your Shipping Method' );  // Title shown in admin
					$this->method_description = __( 'Description of your shipping method' ); // Description shown in admin

					$this->enabled            = "yes"; // This can be added as an setting but for this example its forced enabled
					$this->title              = "My Shipping Method"; // This can be added as an setting but for this example its forced.

					$this->init();
				}

				/**
				 * Init your settings
				 *
				 * @access public
				 * @return void
				 */
				function init() {
					// Load the settings API
					$this->init_form_fields(); // This is part of the settings API. Override the method to add your own settings
					$this->init_settings(); // This is part of the settings API. Loads settings you previously init.

					// Save settings in admin if you have any defined
					add_action( 'woocommerce_update_options_shipping_' . $this->id, array( $this, 'process_admin_options' ) );
				}

				/**
				 * calculate_shipping function.
				 *
				 * @access public
				 * @param mixed $package
				 * @return void
				 */
				public function calculate_shipping( $package ) {
					$rate = array(
						'id' => $this->id,
						'label' => $this->title,
						'cost' => '10.99',
						'calc_tax' => 'per_item'
					);

					// Register the rate
					$this->add_rate( $rate );
				}
				
				/**
				 * Initialize integration settings form fields.
				 *
				 * @return void
				 */
				public function init_form_fields() {
					$log_path = SS_SHIPPING_WC()->get_log_url();
					$agents_address_format = SS_SHIPPING_WC()->get_agents_address_format();

					$this->form_fields = array(
						'test'            	=> array(
							'title'           	=> __( 'Test setting' ),
							'type'            	=> 'text',
							'default'           => '',
							'description'     	=> __( 'Test description' ),
							'desc_tip'        	=> false
						)
					);
				}
				
				public function validate_test_field($key, $value ) {
    					WC_Admin_Settings::add_error("TestError: Key:{$key}::Value:{$value}");//Used for debugging
    					throw new Exception('Invalid value: {$value}');
				}

			}
		}
	}

	add_action( 'woocommerce_shipping_init', 'your_shipping_method_init' );

	function add_your_shipping_method( $methods ) {
		$methods['your_shipping_method'] = 'WC_Your_Shipping_Method';
		return $methods;
	}

	add_filter( 'woocommerce_shipping_methods', 'add_your_shipping_method' );
```

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

Home URL: http://wordpress:8888
Site URL: http://wordpress:8888
WC Version: 3.5.0
Log Directory Writable: ✔
WP Version: 5.0-alpha-43661
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 5.6.31 - We recommend using PHP version 7.2 or above for greater performance and security. How to update your PHP version
PHP Post Max Size: 32 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.35
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.5.0
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 3.31MB
Database Data Size: 2.18MB
Database Index Size: 1.13MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.06MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.06MB + Index: 0.03MB
wp_comments: Data: 0.09MB + Index: 0.09MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_nextend2_image_storage: Data: 0.02MB + Index: 0.02MB
wp_nextend2_section_storage: Data: 0.02MB + Index: 0.03MB
wp_nextend2_smartslider3_generators: Data: 0.02MB + Index: 0.00MB
wp_nextend2_smartslider3_sliders: Data: 0.02MB + Index: 0.00MB
wp_nextend2_smartslider3_sliders_xref: Data: 0.02MB + Index: 0.00MB
wp_nextend2_smartslider3_slides: Data: 0.02MB + Index: 0.00MB
wp_options: Data: 1.03MB + Index: 0.05MB
wp_postmeta: Data: 0.30MB + Index: 0.22MB
wp_posts: Data: 0.08MB + Index: 0.06MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.02MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB

### Post Type Counts ###

attachment: 51
nav_menu_item: 12
page: 6
post: 3
product: 24
product_variation: 4
revision: 5
shipping_package: 2
shop_order: 24

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (8) ###

Your Shipping plugin: by WooThemes – 1.0.0
WooCommerce Beta Tester: by WooCommerce – 2.0.0
WooCommerce Sequential Order Numbers: by SkyVerge – 1.8.3 – Not tested with the active version of WooCommerce
WooCommerce Shipment Tracking: by WooCommerce – 1.6.8 – 1.6.12 is available – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.5.0-rc.1
WordPress Beta Tester: by Peter Westwood – 1.2.3
WordPress Importer: by wordpressdotorg – 0.6.4
WP Log Viewer: by Maxwell Berkel – 1.2.1

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: DKK (DKK)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.3.1 – 2.3.3 is available
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

**Comment by claudiulodro** [comment]
The double notice problem is fixed in #21494.

The issue where notices are not displayed when an exception is thrown in your example code is not an issue (things are working correctly). If you look at [`WC_Settings_API::process_admin_options`](https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-settings-api.php#L214), the error gets set to the shipping method object and not sent to `WC_Admin_Settings` when an exeption is thrown. You'll have to call the [`display_errors`](https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-settings-api.php#L241) function of the custom shipping method object to see those errors if you want them displayed.

**Comment by bilfeldt** [comment]
@claudiulodro thanks for the update, and for #21494.

Regarding validation exception and how to display the error messages. What would be the correct approach to implement validation where the field is NOT saved if it fails and the error is shown on the settings page afterwards?

Maybe it would be nice to add this to the documentation for the [WooCommerce Settings AP](https://docs.woocommerce.com/document/settings-api/)I which is referenced in the [WooCommerce Shipping API](https://docs.woocommerce.com/document/shipping-method-api/#section-4).
